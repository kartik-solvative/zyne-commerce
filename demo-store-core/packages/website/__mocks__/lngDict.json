{
  "general": {
    "viewAll": "View all",
    "search": "search",
    "addToCart": "Add to cart",
    "chooseCountry": "Choose country",
    "international": "International",
    "resultsFor": "{{count}} results for \"{{query}}\""
  },
  "seo": {
    "title": "Commerce Layer Store",
    "description": "Commerce Layer is a transactional commerce API that lets you add multi-market ecommerce to any digital experience, with ease."
  },
  "search": {
    "variants": {
    "color": "Colors",
    "size": "Choose size"
    },
    "facets": {
      "gender": "Gender",
      "category": "Category",
      "price": {
        "amount_float": "Price"
      },
      "available": "Show only available",
      "color": "Color",
      "size": "Size"
    },
    "values": {
      "#000000": "Black",
      "#FFFFFF": "White",
      "#C2C2C2": "Grey",
      "#E7DDC7": "Beige"
    }
  },
  "product": {
    "description": "Product info"
  }
}
