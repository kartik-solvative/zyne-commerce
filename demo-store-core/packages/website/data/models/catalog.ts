import { getRawDataCatalogs } from '#data/catalogs'
import { getRawDataTaxonomies } from '#data/taxonomies'
import { getRawDataTaxons } from '#data/taxons'
import type { Locale } from '#i18n/locale'
import { translateField } from '#utils/locale'
import { LocalizedProductWithVariant, spreadProductVariants } from '#utils/products'
import { makeUnserializable, Unserializable } from '#utils/unserializable'
import { isNotNullish } from '#utils/utility-types'
import type { RawDataCatalog, RawDataTaxon, RawDataTaxonomy } from '@commercelayer/demo-store-types'
import { memoize } from '#utils/memoize'
import uniq from 'lodash/uniq'

export type Catalog = Omit<RawDataCatalog, 'taxonomies'> & {
  taxonomies: Taxonomy[]
}

export type Taxonomy = Omit<RawDataTaxonomy, 'taxons'> & {
  taxons: Taxon[]
}

export type Taxon = Omit<RawDataTaxon, 'label' | 'description' | 'taxons'> & {
  label: string
  description: string
  taxons: Taxon[]
}

export const getCatalog = memoize(
  async (locale: Locale): Promise<Unserializable<Catalog>> => {
    const name = locale.country?.catalog || locale.language.catalog
    const rawDataCatalog = (await getRawDataCatalogs()).value.find(catalog => catalog.name === name)

    if (!rawDataCatalog) {
      throw new Error(`Cannot find the catalog with name "${name}"`)
    }

    return resolveCatalog(rawDataCatalog, locale.code, {})
  },
  locale => locale.code
)

const resolveCatalog = async (catalog: RawDataCatalog, locale: string, productDataset: ProductDataset): Promise<Unserializable<Catalog>> => {
  const taxonomies = await Promise.all(
    catalog.taxonomies
      .map(async taxonomy => resolveTaxonomy(await getTaxonomy(taxonomy), locale, Object.values(productDataset)))
  )

  return makeUnserializable({
    id: catalog.id,
    name: catalog.name,
    taxonomies
  })
}

const getTaxonomy = memoize(
  async (id: string): Promise<RawDataTaxonomy> => {
    const taxonomy = (await getRawDataTaxonomies()).value.find(taxonomy => taxonomy.id === id)

    if (!taxonomy) {
      throw new Error(`Cannot find taxonomy with id ${id}`)
    }

    return taxonomy
  }
)

const resolveTaxonomy = async (taxonomy: RawDataTaxonomy, locale: string, productList: LocalizedProductWithVariant[]): Promise<Taxonomy> => {
  const taxons = await Promise.all(
    taxonomy.taxons
      .map(async taxon => resolveTaxon(await getTaxon(taxon), locale, productList))
  )

  return {
    id: taxonomy.id,
    facetKey: taxonomy.facetKey,
    name: taxonomy.name,
    taxons
  }
}

const getTaxon = memoize(
  async (id: string): Promise<RawDataTaxon> => {
    const taxon = (await getRawDataTaxons()).value.find(taxon => taxon.id === id)

    if (!taxon) {
      throw new Error(`Cannot find taxon with id ${id}`)
    }

    return taxon
  }
)

const resolveTaxon = async (taxon: RawDataTaxon, locale: string, productList: LocalizedProductWithVariant[]): Promise<Taxon> => {
  return {
    id: taxon.id,
    label: translateField(taxon, 'label', locale),
    description: translateField(taxon, 'description', locale),
    name: taxon.name,
    slug: taxon.slug,
    taxons: await Promise.all(
      taxon.taxons?.map(async taxon => resolveTaxon(await getTaxon(taxon), locale, productList)) || []
    ),
    references: taxon.references
  }
}


// --------------------------------------------------------------------------------------------------------------------

type ProductDataset = {
  [code: LocalizedProductWithVariant['sku']]: LocalizedProductWithVariant
}

export async function buildProductDataset(catalog: RawDataCatalog, locale: string, productList: LocalizedProductWithVariant[] = []): Promise<ProductDataset> {
  if (productList.length <= 0) {
    return {}
  }

  const productDataset: ProductDataset = {}

  for (const taxonomyKey of catalog.taxonomies) {
    const taxonomy = await getTaxonomy(taxonomyKey)

    for (const taxonKey of taxonomy.taxons) {
      await buildProductDataset_taxon(taxonKey, [], taxonomy, locale, productList)
    }
  }

  async function buildProductDataset_taxon(taxonKey: string, prevTaxons: RawDataTaxon[], taxonomy: RawDataTaxonomy, locale: string, productList: LocalizedProductWithVariant[]): Promise<void> {
    const taxon = await getTaxon(taxonKey)

    const products = (await flattenReferences(taxon))
      .map(ref => productList.find(product => product.sku === ref))
      .filter(isNotNullish)

    spreadProductVariants(products).forEach(product => {
      productDataset[product.sku] = productDataset[product.sku] || product
      productDataset[product.sku] = {
        ...productDataset[product.sku],
        [taxonomy.facetKey]: uniq([
          // @ts-expect-error
          ...(productDataset[product.sku][taxonomy.facetKey] || []),
          `${prevTaxons.length > 0 ? prevTaxons.map(t => `${translateField(t, 'label', locale)} > `).join('') : ''}${translateField(taxon, 'label', locale)}`
        ])
      }
    })

    taxon.taxons?.forEach(taxonKey => buildProductDataset_taxon(taxonKey, prevTaxons.concat(taxon), taxonomy, locale, productList))
  }

  return productDataset
}

async function flattenReferences(taxon: RawDataTaxon): Promise<string[]> {
  const subReferences = (await Promise.all(
    (taxon.taxons || []).flatMap(
      async taxonKey => {
        const taxon = await getTaxon(taxonKey)
        return await flattenReferences(taxon)
      }
    )
  )).flat()

  return taxon.references.concat(subReferences)
}
