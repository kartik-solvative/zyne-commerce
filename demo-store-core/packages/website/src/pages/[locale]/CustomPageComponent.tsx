import type { HeaderProps } from '#components/Header'
import { Page } from '#components/Page'
import { CarouselPageComponent } from '#components/PageComponents/CarouselPageComponent'
import { GridPageComponent } from '#components/PageComponents/GridPageComponent'
import { HeroPageComponent } from '#components/PageComponents/HeroPageComponent'
import { MarkdownPageComponent } from '#components/PageComponents/MarkdownPageComponent'
import { ProductGridPageComponent } from '#components/PageComponents/ProductGridPageComponent'
import { Title } from '#components/Title'
import type { CustomPage } from '#utils/pages'
import type { NextPage } from 'next'

export type Props = HeaderProps & {
  page: CustomPage
  localeCodes: string[]
}

export const CustomPageComponent: NextPage<Props> = ({ navigation, page, localeCodes }) => {
  console.log(page,"page");
  return (
    <Page localeCodes={localeCodes} navigation={navigation} title={page.title || undefined} description={page.description || undefined}>
      <div data-testid='page-components' className='mt-8 flex flex-col gap-8'>
        {page.title  && <Title title={page.title}></Title>}

        {
          page.components.map(component => {
            switch (component.type) {

              case 'carousel':
                return <CarouselPageComponent key={component.id} component={component} />

              case 'hero':
                return <HeroPageComponent key={component.id} component={component} />

              case 'grid':
                return <GridPageComponent key={component.id} component={component} />

              case 'product-grid':
                return <ProductGridPageComponent key={component.id} component={component} />

              case 'markdown':
                return <MarkdownPageComponent key={component.id} component={component} />

              default:
                ((_: never): void => { })(component)
                break;
            }
          })
        }
      </div>
    </Page>
  )
}
