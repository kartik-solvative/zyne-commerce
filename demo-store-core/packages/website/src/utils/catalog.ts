import type { Catalog, Taxon, Taxonomy } from '#data/models/catalog'
import { deepFind, DeepFindResult } from '#utils/collection'
import { getSearchUrl } from '#utils/url'
import uniq from 'lodash/uniq'
import type { Unserializable } from './unserializable'

export type NavigationLink = {
  key: string
  href: string
  text: string
  description: string
  children?: NavigationLink[]
}

export type Breadcrumbs = {
  parent: NavigationLink[]
  current: NavigationLink
  children: NavigationLink[]
}

export type Navigation = {
  current: NavigationLink
  path: NavigationLink[]
}


const getPrimaryTaxonomy = (catalog: Unserializable<Catalog>): Taxonomy => {
  // TODO: taxonomies[0] is a requirement. First taxonomy is considered the navigation one. Should it be configurable?
  return catalog.value.taxonomies[0]
}

export const getRootNavigationLinks = (catalog: Unserializable<Catalog>): NavigationLink[] => {
  return getPrimaryTaxonomy(catalog).taxons.map(taxon => ({
    key: taxon.id,
    href: getSearchUrl(taxon),
    text: taxon.label,
    description: taxon.description
  }))
}

export const getBreadcrumbs = (taxon: DeepFindResult<Taxon>): Breadcrumbs => {
  return {
    parent: taxon.memo.map(({ id: key, slug, label, description }) => ({
      key: key,
      href: getSearchUrl({slug}),
      text: label,
      description
    })),
    current: {
      key: taxon.result.id,
      href: getSearchUrl(taxon.result),
      text: taxon.result.label,
      description: taxon.result.description
    },
    children: taxon.result.taxons.map(({ id: key, slug, label, description }) => ({
      key: key,
      href: getSearchUrl({slug}),
      text: label,
      description
    }))
  }
}

const getNavigationChildren = (foundTaxon: DeepFindResult<Taxon>, index: number, currentTaxon: Taxon): NavigationLink[] => {
  return foundTaxon.memo[index]?.id === currentTaxon.id ? currentTaxon.taxons.filter(t => flattenReferencesFromTaxon(t).length > 0).map((taxon) => ({
    key: taxon.id,
    href: getSearchUrl({ slug: taxon.slug }),
    text: taxon.label,
    description: taxon.description,
    children: getNavigationChildren(foundTaxon, ++index, taxon)
  })) : []
}

export const getNavigation = (foundTaxon: DeepFindResult<Taxon>): Navigation => {
  return {
    current: {
      key: foundTaxon.result.id,
      href: getSearchUrl(foundTaxon.result),
      text: foundTaxon.result.label,
      description: foundTaxon.result.description
    },
    path: [{
      key: foundTaxon.memo[0].id,
      href: getSearchUrl(foundTaxon.memo[0]),
      text: foundTaxon.memo[0].label,
      description: foundTaxon.memo[0].description,
      children: getNavigationChildren(foundTaxon, 0, foundTaxon.memo[0])
    }]
  }
}

export const getSlugs = (catalog: Unserializable<Catalog>): string[] => {
  function getFlatSlug(taxon: Taxon): string[] {
    return [taxon.slug].concat(taxon.taxons.flatMap(getFlatSlug))
  }

  return getPrimaryTaxonomy(catalog).taxons.flatMap(getFlatSlug)
}

export function flattenReferencesFromTaxon(taxon: Taxon): string[] {
  return uniq(
    taxon.references.concat(taxon.taxons?.flatMap(flattenReferencesFromTaxon) || [])
  )
}

export function flattenReferencesFromCatalog(catalog: Unserializable<Catalog>): string[] {
  return uniq(
    catalog.value.taxonomies.flatMap(({ taxons }) => taxons.flatMap(flattenReferencesFromTaxon)),
  )
}

export function findTaxonBySlug(catalog: Unserializable<Catalog>, slug: string): DeepFindResult<Taxon> {
  const taxon = catalog.value.taxonomies.reduce((acc, cv) => {
    if (acc) {
      return acc
    }

    return deepFind(cv.taxons, 'taxons', 'slug', slug)
  }, undefined as DeepFindResult<Taxon> | undefined)

  if (!taxon) {
    throw new Error('Cannot find Taxon!')
  }

  return taxon
}
