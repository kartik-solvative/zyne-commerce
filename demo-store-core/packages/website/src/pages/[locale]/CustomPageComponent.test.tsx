import { getRootNavigationLinks } from '#utils/catalog'
import type { CustomPage } from '#utils/pages'
import lngDict from '#__mocks__/lngDict.json'
import { act, render, screen } from '@testing-library/react'
import { createCarouselPageComponent, createCatalog, createRouter, createTaxon, createTaxonomy } from 'jest.helpers'
import { I18nProvider } from 'next-localization'
import { CustomPageComponent } from './CustomPageComponent'

jest.mock('#data/languages', () => ({
  getRawDataLanguages: () => ([
    {
      "name": "English",
      "code": "en",
      "catalog": "AMER"
    },
    {
      "name": "Italiano",
      "code": "it",
      "catalog": "AMER"
    }
  ])
}))

const useRouter = jest.spyOn(require('next/router'), 'useRouter')

beforeEach(() => {
  jest.useFakeTimers()

  useRouter.mockReset()
})

afterEach(() => {
  jest.clearAllTimers()
});

test('CustomPage component', async () => {
  useRouter.mockImplementation(() => createRouter('/'))

  const taxon = createTaxon('1')
  const taxonomy = createTaxonomy('1', [taxon])
  const catalog = createCatalog('1', [taxonomy])
  const navigation = getRootNavigationLinks(catalog)

  const carouselPageComponent = createCarouselPageComponent('1')

  const page: CustomPage = {
    slug: '/',
    title: 'Page title',
    description: 'Page description',
    components: [carouselPageComponent]
  }

  let container
  await act(async () => {
    ({ container } = render(
      <I18nProvider lngDict={lngDict} locale='en'>
        <CustomPageComponent localeCodes={['en', 'en-US', 'it-IT']} navigation={navigation} page={page} />
      </I18nProvider>
    ))

    jest.runAllTimers();
  })

  expect(container).toMatchSnapshot()

  const pageComponents = await screen.findAllByTestId('carousel-page-component')
  expect(pageComponents.length).toStrictEqual(1)

  const title = await screen.findByRole('heading', { level: 1 })
  expect(title).toHaveTextContent('Page title')
})
