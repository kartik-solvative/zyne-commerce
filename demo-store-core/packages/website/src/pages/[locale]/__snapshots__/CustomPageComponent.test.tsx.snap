// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CustomPage component 1`] = `
<div>
  <main
    class="flex flex-col h-screen"
  >
    <div
      class="container mx-auto px-6"
    >
      <header
        class="border-b-gray-200 border-b pb-6 sticky top-0 bg-pageBackground z-50"
      >
        <nav
          class="flex items-center justify-between flex-wrap mb-4"
        >
          <div
            class="flex items-center flex-no-shrink text-white mr-6"
          >
            <a
              href="/en-US"
            >
              <div
                class="text-black py-6 pb-6 "
              >
                <svg
                  aria-label="Commerce Layer logomark"
                  class="m-1 inline"
                  width="32"
                />
                <svg
                  aria-label="Commerce Layer logotype"
                  class="ml-2 hidden md:inline"
                  width="192"
                />
              </div>
            </a>
          </div>
          <div
            class="flex items-center w-auto flex-grow justify-end"
          />
        </nav>
        <div
          class="flex items-center justify-between flex-wrap gap-4 relative"
        >
          <div
            class="order-2 lg:order-1"
          >
            <nav
              class="hidden lg:block lg:relative absolute right-0 top-full w-full z-10 flex-col bg-pageBackground py-4 lg:py-0"
            >
              <a
                class="text-back font-bold py-2 px-3 my-2 mx-1 transition-colors duration-300 bg-transparent"
                href="/en-US/search/taxon-slug-1"
              >
                Label for 1
              </a>
            </nav>
            <button
              aria-controls="menu"
              aria-label="Open menu"
              class="w-11 h-11 flex flex-col items-center justify-center space-y-1 bg-black rounded block lg:hidden"
              type="button"
            >
              <span
                class="w-4 h-0.5 bg-white"
              />
              <span
                class="w-4 h-0.5 bg-white"
              />
              <span
                class="w-4 h-0.5 bg-white"
              />
            </button>
          </div>
          <div
            class="order-1 grow lg:grow-0"
          >
            <label
              class="relative py-2.5 rounded bg-gray-100 text-gray-400 focus-within:text-gray-600 block"
              for="email"
            >
              <svg
                class="pointer-events-none w-6 h-6 absolute top-1/2 transform -translate-y-1/2 left-3"
              />
              <form
                action="/en-US/search"
              >
                <input
                  class="form-input appearance-none bg-transparent w-full pl-14 focus:outline-none focus:shadow-outline"
                  placeholder="search"
                  value=""
                />
              </form>
            </label>
          </div>
        </div>
      </header>
      <div
        class="mt-8 flex flex-col gap-8"
        data-testid="page-components"
      >
        <div
          class="flex items-center py-8 relative border-b-gray-200 border-b mb-10"
        >
          <h1
            class="text-2xl flex-auto font-semibold text-black py-1"
          >
            Page title
          </h1>
        </div>
        <div
          data-testid="carousel-page-component"
        >
          <div
            class="swiper swiper-initialized swiper-horizontal carousel"
          >
            <div
              class="swiper-wrapper"
              style="transition-duration: 0ms; transform: translate3d(NaNpx, 0px, 0px);"
            >
              <div
                class="swiper-slide select-none"
              >
                <div
                  class="relative rounded-lg overflow-hidden w-full max-h-[70vh] min-h-[24rem]"
                >
                  <img
                    alt="slide-1-image-alt-1"
                    class="object-cover object-center w-full rounded-lg"
                    src="slide-1-image-src-1"
                  />
                  <div
                    class="hidden md:block absolute inset-0 bg-gradient-to-b from-black to-black/0 opacity-50"
                  />
                  <div
                    class="flex justify-center relative -top-4 md:text-white md:absolute md:pr-10 md:top-1/2 md:right-0 md:w-1/2 md:-translate-y-1/2"
                  >
                    <div
                      class="bg-white w-11/12 rounded-md flex justify-center py-4 md:bg-transparent"
                    >
                      <div
                        class="w-11/12"
                      >
                        <div
                          class="font-bold text-xl md:text-3xl mb-4"
                        >
                          slide-1-title-1
                        </div>
                        <div
                          class="leading-snug mb-4"
                        >
                          slide-1-description-1
                        </div>
                        <a
                          class="py-2 px-4 inline-block font-semibold rounded-md bg-violet-400 text-white text-sm w-full text-center md:w-auto"
                          href="/en-USslide-1-linkHref-1"
                        >
                          slide-1-linkLabel-1
                        </a>
                      </div>
                    </div>
                  </div>
                </div>
              </div>
              <div
                class="swiper-slide select-none"
              >
                <div
                  class="relative rounded-lg overflow-hidden w-full max-h-[70vh] min-h-[24rem]"
                >
                  <img
                    alt="slide-2-image-alt-1"
                    class="object-cover object-center w-full rounded-lg"
                    src="slide-2-image-src-1"
                  />
                  <div
                    class="hidden md:block absolute inset-0 bg-gradient-to-b from-black to-black/0 opacity-50"
                  />
                  <div
                    class="flex justify-center relative -top-4 md:text-white md:absolute md:pr-10 md:top-1/2 md:right-0 md:w-1/2 md:-translate-y-1/2"
                  >
                    <div
                      class="bg-white w-11/12 rounded-md flex justify-center py-4 md:bg-transparent"
                    >
                      <div
                        class="w-11/12"
                      >
                        <div
                          class="font-bold text-xl md:text-3xl mb-4"
                        >
                          slide-2-title-1
                        </div>
                        <div
                          class="leading-snug mb-4"
                        >
                          slide-2-description-1
                        </div>
                        <a
                          class="py-2 px-4 inline-block font-semibold rounded-md bg-violet-400 text-white text-sm w-full text-center md:w-auto"
                          href="/en-USslide-2-linkHref-1"
                        >
                          slide-2-linkLabel-1
                        </a>
                      </div>
                    </div>
                  </div>
                </div>
              </div>
            </div>
            <div
              class="swiper-button-prev"
            />
            <div
              class="swiper-button-next"
            />
            <div
              class="swiper-pagination swiper-pagination-bullets swiper-pagination-horizontal"
            />
          </div>
        </div>
      </div>
    </div>
    <div
      class="flex-grow py-8 lg:p-16 mt-24 bg-gray-50"
    >
      <div
        class="container mx-auto px-6"
      >
        <div
          class="lg:flex items-center justify-between flex-wrap"
        >
          <div>
            <div
              class="flex items-center justify-between"
            >
              <span
                class="flex items-center"
              >
                <svg
                  class="mr-1.5"
                />
                 
                International
              </span>
              <a
                class="uppercase text-violet-400 text-xs border-b border-gray-200 ml-3 mt-1"
                href="/"
              >
                Choose country
              </a>
            </div>
          </div>
          <div
            class="mt-6 lg:mt-0"
          >
            <select
              class=" inputSelect"
            >
              <option
                selected=""
                value="en"
              >
                English
              </option>
              <option
                value="it"
              >
                Italiano
              </option>
            </select>
          </div>
        </div>
        <div
          class="mt-6 py-6 border-y border-y-gray-100 text-xs flex flex-col md:flex-row md:divide-x"
        >
          <div
            class="py-2 md:py-0 md:pr-4"
          >
            <a
              href="/en-US/shipping-and-payments"
            >
              Shipping & Payments
            </a>
          </div>
          <div
            class="py-2 md:py-0 md:px-4"
          >
            <a
              href="/en-US/returns"
            >
              Returns
            </a>
          </div>
          <div
            class="py-2 md:py-0 md:px-4"
          >
            <a
              href="/en-US/terms-and-conditions"
            >
              Terms and conditions
            </a>
          </div>
          <div
            class="py-2 md:py-0 md:px-4"
          >
            <a
              href="/en-US/cookies-and-privacy"
            >
              Cookies and privacy
            </a>
          </div>
        </div>
        <div
          class="mt-12 flex justify-between items-center flex-col gap-6 md:flex-row"
        >
          <div
            class="text-gray-500 flex gap-4 items-start"
          >
            <svg
              aria-label="Commerce Layer logomark"
              width="32"
            />
            <div
              class="text-xs leading-relaxed"
            >
              Copyright © 2022. Commerce Layer, Inc.
              <br />
              2965 Woodside Road, Woodside, CA 94062
              <br />
            </div>
          </div>
          <div
            class="text-gray-500 flex gap-4 items-start grayscale"
          >
            <svg />
            <svg />
            <svg />
            <svg />
            <svg />
            <svg />
          </div>
        </div>
      </div>
    </div>
  </main>
</div>
`;
